internal/terraform/node_resource_plan_partialexp.go

```diff
@@ -302,13 +282,96 @@ func (n *nodePlannablePartialExpandedResource) managedResourceExecute(ctx EvalCo
 	return &change, diags
 }
 
-func (n *nodePlannablePartialExpandedResource) dataResourceExecute(ctx EvalContext) (cty.Value, tfdiags.Diagnostics) {
+// Logic here mirrors a combination of (*NodePlannableResourceInstance).dataResourceExecute
+// and (*NodeAbstractResourceInstance).planDataSource.
+func (n *nodePlannablePartialExpandedResource) dataResourceExecute(ctx EvalContext) (*plans.ResourceInstanceChange, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 
-	// TODO: Ideally we should do an approximation of the normal data resource
-	// planning process similar to what we're doing for managed resources in
-	// managedResourceExecute, but we'll save that for a later phase of this
-	// experiment since managed resources are enough to start getting real
-	// experience with this new evaluation approach.
-	return cty.DynamicVal, diags
+	// Start with a basic change, then attempt to fill in the After value.
+	change := plans.ResourceInstanceChange{
+		Addr:         n.addr.UnknownResourceInstance(),
+		ProviderAddr: n.resolvedProvider,
+		Change: plans.Change{
+			// Data sources can only Read.
+			Action: plans.Read,
+			Before: cty.NullVal(cty.DynamicPseudoType),
+			After:  cty.DynamicVal, // hoping to fill this in
+		},
+		// For now, this is the default reason for deferred data source reads.
+		// It's _basically_ the truth!
+		ActionReason: plans.ResourceInstanceReadBecauseConfigUnknown,
+	}
+
+	// Unlike with the managed path, we don't ask the provider to *do* anything.
+	_, providerSchema, err := getProvider(ctx, n.resolvedProvider)
+	diags = diags.Append(err)
+	if diags.HasErrors() {
+		return &change, diags
+	}
+
+	diags = diags.Append(validateSelfRef(n.addr.Resource(), n.config.Config, providerSchema))
+	if diags.HasErrors() {
+		return &change, diags
+	}
+
+	// This is the point where we switch to mirroring logic from
+	// NodeAbstractResourceInstance's planDataSource. If you were curious.
+
+	schema, _ := providerSchema.SchemaForResourceAddr(n.addr.Resource())
+	if schema == nil {
+		// Should be caught during validation, so we don't bother with a pretty error here
+		diags = diags.Append(fmt.Errorf("provider does not support resource type %q", n.addr.Resource().Type))
+		return &change, diags
+	}
+
+	keyData := n.keyData()
+
+	configVal, _, configDiags := ctx.EvaluateBlock(n.config.Config, schema, nil, keyData)
+	diags = diags.Append(configDiags)
+	if configDiags.HasErrors() {
+		return &change, diags
+	}
+
+	// TODO NF: Do we need to do something here to handle partial expanded data
```

Comment:
I *think* nested data sources don't support the for each attribute so I'd say no.

Reply:
Oh hmm just realized they might be in a module that has an unknown for_each, so the possibility exists... but I re-read the special case handling and it's fine, because it's all around "when do we perform the read, and what do we do with the result?", none of which is actually relevant if you're deferred anyway. I'll update the comment with that.